spec/controllers/api/v0/shipments_controller_spec.rb

```diff
@@ -110,28 +110,69 @@
       expect(response.status).to eq(422)
     end
 
-    context 'for completed shipments' do
+    describe "#add and #remove" do
       let(:order) { create :completed_order_with_totals }
+      let(:line_item) { order.line_items.first }
+      let(:existing_variant) { line_item.variant }
+      let(:new_variant) { create(:variant) }
+      let(:params) {
+        {
+          quantity: 2,
+          order_id: order.to_param,
+          id: order.shipments.first.to_param
+        }
+      }
 
-      it 'adds a variant to a shipment' do
-        api_put :add, variant_id: variant.to_param,
-                      quantity: 2,
-                      order_id: order.to_param,
-                      id: order.shipments.first.to_param
+      before do
+        line_item.update!(quantity: 3)
+      end
+
+      context 'for completed shipments' do
+        it 'adds a variant to a shipment' do
+          expect {
+            api_put :add, params.merge(variant_id: new_variant.to_param)
+            expect(response.status).to eq(200)
+          }.to change { inventory_units_for(new_variant).size }.by(2)
+        end
+
+        it 'adjusts stock when adding a variant' do
+          expect {
+            api_put :add, params.merge(variant_id: new_variant.to_param)
+          }.to change { new_variant.reload.on_hand }.by(-2)
+        end
+
+        it 'removes a variant from a shipment' do
+          expect {
+            api_put :remove, params.merge(variant_id: existing_variant.to_param)
+            expect(response.status).to eq(200)
+          }.to change { inventory_units_for(existing_variant).size }.by(-2)
+        end
 
-        expect(response.status).to eq(200)
-        expect(inventory_units_for(variant).size).to eq 2
+        it 'adjusts stock when removing a variant' do
+          expect {
+            api_put :remove, params.merge(variant_id: existing_variant.to_param)
+          }.to change { existing_variant.reload.on_hand }.by(2)
+        end
       end
 
-      it 'removes a variant from a shipment' do
-        order.contents.add(variant, 2)
-        api_put :remove, variant_id: variant.to_param,
-                         quantity: 1,
-                         order_id: order.to_param,
-                         id: order.shipments.first.to_param
+      context "for canceled orders" do
+        before do
+          expect(order.cancel).to eq true
```

Comment:
I'm not a big fan of asserting things in the test setup phase. It feels a bit like defensive and unconfident programming and it adds up to the build time. If `cancel` didn't work we would notice because the test would fail, right?

Reply:
I'm not a big fan of asserting in the setup either. But I do want to know if something goes wrong here and I don't think that this assertion slows the tests down much.

If `cancel` doesn't work then the order might be in an unknown state with unknown behaviour and we don't know if that would pass or fail the test. I wish I could rely on it to work. My experience is though that it can be tricky to get the test setup right and then things don't behave as expected.

The best solution would probably be to add a new factory: `create(:cancelled_order)`. Another time...